src/rust/engine/process_execution/src/cache.rs

```diff
@@ -0,0 +1,181 @@
+
+#[derive(Clone)]
+struct CommandRunner {
+  underlying: Arc<dyn crate::CommandRunner>,
+  process_execution_store: ShardedLmdb,
+  file_store: Store,
+}
+
+impl crate::CommandRunner for CommandRunner {
+  fn run(&self, req: ExecuteProcessRequest) -> BoxFuture<FallibleExecuteProcessResult, String> {
+    let digest = try_future!(self.digest(&req));
+    let key = digest.0;
+
+    let command_runner = self.clone();
+    self
+      .lookup(key)
+      .and_then(move |maybe_result| {
+        if let Some(result) = maybe_result {
+          futures::future::ok(result).to_boxed()
+        } else {
+          command_runner
+            .underlying
+            .run(req)
+            .and_then(move |result| command_runner.store(key, &result).map(|()| result))
+            .to_boxed()
+        }
+      })
+      .to_boxed()
+  }
+}
+
+impl CommandRunner {
+  fn digest(&self, req: &ExecuteProcessRequest) -> Result<Digest, String> {
+    // TODO: Parameterise options
+    let (_action, _command, execute_request) =
+      crate::remote::make_execute_request(req, &None, &None, BTreeMap::new())?;
+    execute_request.get_action_digest().into()
+  }
+
+  fn lookup(
+    &self,
+    fingerprint: Fingerprint,
+  ) -> impl Future<Item = Option<FallibleExecuteProcessResult>, Error = String> {
+    let file_store = self.file_store.clone();
+    self
+      .process_execution_store
+      .load_bytes_with(fingerprint, |bytes| {
+        let mut execute_response = bazel_protos::remote_execution::ExecuteResponse::new();
+        execute_response
+          .merge_from_bytes(&bytes)
+          .map_err(|e| format!("Invalid ExecuteResponse: {:?}", e))?;
+        Ok(execute_response)
+      })
+      .and_then(move |maybe_execute_response| {
+        if let Some(execute_response) = maybe_execute_response {
+          crate::remote::populate_fallible_execution_result(file_store, execute_response, vec![])
+            .map(Some)
+            .to_boxed()
+        } else {
+          futures::future::ok(None).to_boxed()
+        }
+      })
+  }
+
+  fn store(
+    &self,
+    fingerprint: Fingerprint,
+    result: &FallibleExecuteProcessResult,
+  ) -> impl Future<Item = (), Error = String> {
+    let mut execute_response = bazel_protos::remote_execution::ExecuteResponse::new();
+    execute_response.set_cached_result(true);
+    let action_result = execute_response.mut_result();
+    action_result.set_exit_code(result.exit_code);
+    action_result.mut_output_directories().push({
+      let mut directory = bazel_protos::remote_execution::OutputDirectory::new();
+      directory.set_path(String::new());
+      directory.set_tree_digest((&result.output_directory).into());
+      directory
+    });
+    let process_execution_store = self.process_execution_store.clone();
+    self
+      .file_store
+      .store_file_bytes(result.stdout.clone(), true)
+      .join(
+        self
+          .file_store
+          .store_file_bytes(result.stderr.clone(), true),
```

Comment:
If we fail to get any of these, we should probably just treat it as a cache miss rather than a failure (possibly with `debug` or `trace` logging), because I can definitely imagine some of the dependency data expiring out here a little bit out of order with the actual cache entry.

Reply:
Done